File: CAN_DRIVER/NANOTEC/NanoLib_1.0.1/NanoLib C# Windows/nanolib_csharp_win_1.0.1/Example/NanolibExample/NanolibExample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NanolibCsharpExampleWin
{
    internal class NanolibExample
    {
        private static void Main(string[] args)
        {
            NanolibHelper nanolibHelper = new NanolibHelper();


            try
            {
                nanolibHelper.Setup();

                // its possible to set the logging level to a different level
                nanolibHelper.SetLoggingLevel(Nlc.LogLevel.Off);

                // list all hardware available, decide for the first one
                Nlc.BusHWIdVector busHwIds = nanolibHelper.GetBusHardware();

                if (busHwIds.Count() <= 0)
                {
                    Console.Error.WriteLine("No bus hardware found");

                    //Wait before close console
                    PressAnyKey();
                    return;
                }

                Console.WriteLine();
                Console.WriteLine("Available bus hardware:");
                Console.WriteLine();

                uint lineNum = 0;

                // just for better overview: print out available hardware
                foreach (Nlc.BusHardwareId adapter in busHwIds)
                {
                    Console.Write("{0}. {1}", lineNum, adapter.getName());
                    Console.WriteLine(" with protocol: {0}", adapter.getProtocol());
                    lineNum++;
                }

                Console.WriteLine();

                if (lineNum > 1)
                {
                    Console.Write("Please select (enter) bus hardware number(0-" + (lineNum - 1) + ") and press [ENTER]:");

                    try
                    {
                        lineNum = uint.Parse(Console.ReadLine());
                    }
                    catch (FormatException)
                    {
                    }

                    Console.WriteLine();

                    if (lineNum >= busHwIds.Count)
                    {
                        Console.WriteLine("Invalid selection!");
                        return;
                    }
                }
                else
                {
                    lineNum = 0;
                }

                // let's select the first one
                //Nlc.BusHardwareId busHwId = busHwIds[(int)lineNum]; // Do not do this !!!!
                //Nlc.BusHardwareId busHwId = new Nlc.BusHardwareId(busHwIds[(int)lineNum]); // Very important step in the C# version of NanoLib

                // Create a copy of every object, which is returned in a vector, because
                // when the vector goes out of scope, the contained objects will be destroyed.
                // or just copy the vector into an array
                Nlc.BusHardwareId busHwId = busHwIds.ToArray()[lineNum];

                // create bus hardware options for opening the hardware
                Nlc.BusHardwareOptions busHwOptions = nanolibHelper.CreateBusHardwareOptions(busHwId);

                // now able to open the hardware itself
                nanolibHelper.OpenBusHardware(busHwId, busHwOptions);

                try
                {

                    Console.WriteLine("Scanning for devices...");

                    // either scan the whole bus for devices (in case the bus supports scanning)
                    Nlc.DeviceIdVector deviceIds = nanolibHelper.ScanBus(busHwId);

                    Console.WriteLine("");
                    foreach (Nlc.DeviceId id in deviceIds)
                    {
                        Console.WriteLine("Found device: {0}", id.getDeviceId());
                    }

                    if (deviceIds.Count == 0)
                    {
                        Console.WriteLine("No devices found.");
                        PressAnyKey();
                        return;
                    }

                    Console.WriteLine();
                    Console.WriteLine("Available devices:");
                    Console.WriteLine();

                    lineNum = 0;

                    // just for better overview: print out available devices
                    foreach (Nlc.DeviceId id in deviceIds)
                    {
                        Console.Write("{0}. {1}", lineNum, id.getDescription());
                        Console.WriteLine("[device id: {0}, hardware: {1}]", id.getDeviceId(), id.getBusHardwareId().getName());
                        lineNum++;
                    }

                    Console.WriteLine();

                    if (lineNum > 1)
                    {
                        Console.Write("Please select (enter) device number(0-" + (lineNum - 1) + ") and press [ENTER]:");
                        try
                        {
                            lineNum = uint.Parse(Console.ReadLine());
                        }
                        catch (FormatException)
                        {
                        }

                        Console.WriteLine();

                        if (lineNum >= deviceIds.Count)
                        {
                            Console.WriteLine("Invalid selection!");
                            return;
                        }
                    }
                    else
                    {
                        lineNum = 0;
                    }

                    // We can create the device id manually
                    // Nlc.DeviceId deviceId = new Nlc.DeviceId(busHwId, 1, "");
                    // or get first found device

                    // Nlc.DeviceId deviceId = deviceIds[(int)lineNum]; // Do not do this !!!
                    //Nlc.DeviceId deviceId = new Nlc.DeviceId(deviceIds[(int)lineNum]); // Very important step in the C# version of NanoLib

                    // Create a copy of every object, which is returned in a vector, because
                    // when the vector goes out of scope, the contained objects will be destroyed.
                    // or just copy the vector into an array
                    Nlc.DeviceId deviceId = deviceIds.ToArray()[lineNum];


                    // Create a device and get a handle
                    Nlc.DeviceHandle deviceHandle = nanolibHelper.CreateDevice(deviceId);

                    // now connect to the device
                    nanolibHelper.ConnectDevice(deviceHandle);

                    try
                    {

                        // now ready to work with the device, here are some examples on how to access the
                        // object dictionary:
                        ObjectDictionaryAccessExamples(nanolibHelper, deviceHandle);
                        ObjectDictionaryAccessExamplesViaDictionaryInterface(nanolibHelper, deviceHandle);

                        NanolibSamplerExample samplerExample = new NanolibSamplerExample(nanolibHelper, deviceHandle);

                        if (AskTheUser("Do you want to run sampler examples?"))
                        {
                            // Run sampler examples
                            samplerExample.RunSamplerExamples();
                        }

                        NanolibProfinetDcpExample nanolibProfinetDcpExample = new NanolibProfinetDcpExample(nanolibHelper);

                        Console.WriteLine();

                        // ProfinetDCP example
                        nanolibProfinetDcpExample.ProfinetDcpExample(busHwId);
                    }
                    finally
                    {
                        // cleanup and close everything
                        Console.WriteLine("Disconnecting the device.");
                        nanolibHelper.DisconnectDevice(deviceHandle);
                    }
                } 
                finally
                {
                    // close the hardware
                    Console.WriteLine("Closing the hardware bus.");
                    nanolibHelper.CloseBusHardware(busHwId);
                }

                Console.WriteLine("");

                // Wait before close console
                PressAnyKey();
            }
            catch (NanolibException e)
            {
                Console.Error.WriteLine("Error occured: {0}", e.Message);

                //Wait before close console
                PressAnyKey();
            }
        }

        private static void ObjectDictionaryAccessExamples(NanolibHelper nanolibHelper, Nlc.DeviceHandle deviceId)
        {
            // OD example
            Console.WriteLine("");
            Console.WriteLine("OD Example");
            Console.WriteLine("");

            // read status word
            Console.WriteLine("Reading subindex 0 of index 0x6040");
            long statusWord = nanolibHelper.ReadNumber(deviceId, new Nlc.OdIndex(0x6040, 0x00));
            Console.WriteLine("Result: {0}", statusWord);
            Console.WriteLine("");

            // write control word to value "6"
            Console.WriteLine("Motor Stop (0x6040-0)");
            Console.WriteLine("");
            nanolibHelper.WriteNumber(deviceId, 6, new Nlc.OdIndex(0x6040, 0x00), 16);

            // read homepage of nanotec
            Console.WriteLine("Read Nanotec home page string");
            Console.WriteLine("The home page of Nanotec Electronic GmbH & Co. KG is: {0}", nanolibHelper.ReadString(deviceId, new Nlc.OdIndex(0x6505, 0x00)));
            Console.WriteLine("");

            // read error stack
            Console.WriteLine("Read device error stack");
            Nlc.IntVector errorStack = nanolibHelper.ReadArray(deviceId, 0x1003);
            Console.WriteLine("The error stack has {0} elements", errorStack.ElementAt(0));
            Console.WriteLine("");
        }

        private static void ObjectDictionaryAccessExamplesViaDictionaryInterface(NanolibHelper nanolibHelper, Nlc.DeviceHandle deviceHandle)
        {
            Nlc.OdIndex homePageOd = new Nlc.OdIndex(0x6505, 0x00);
            Nlc.OdIndex controlWordOd = new Nlc.OdIndex(0x6040, 0x00);


            // OD example
            Console.WriteLine("");
            Console.WriteLine("OD Example via Dictionary Interface");
            Console.WriteLine("");

            Nlc.ObjectDictionary objectDictionary = nanolibHelper.GetDeviceObjectDictionary(deviceHandle);

            // read status word
            Console.WriteLine("Reading subindex 0 of index 0x6040");
            long statusWord = nanolibHelper.ReadNumber(objectDictionary, new Nlc.OdIndex(0x6040, 0x00));
            Console.WriteLine("Result: {0}", statusWord);
            Console.WriteLine("");

            // write control word to value "6"
            Console.WriteLine("Motor Stop (0x6040-0)");
            Console.WriteLine("");
            nanolibHelper.WriteNumber(objectDictionary, 6, new Nlc.OdIndex(0x6040, 0x00));

            // read homepage of nanotec
            Console.WriteLine("Read Nanotec home page string");
            Console.WriteLine("The home page of Nanotec Electronic GmbH & Co. KG is: {0}", nanolibHelper.ReadString(objectDictionary, homePageOd));
            Console.WriteLine("");

            Nlc.ObjectEntry objectEntry = nanolibHelper.GetObjectEntry(objectDictionary, controlWordOd.Index);
            Console.WriteLine("Some ObjectEntry properties:");
            Console.WriteLine("Object(0x{0:X04}).ObjectCode = {1}", controlWordOd.Index, objectEntry.getObjectCode().ToString());
            Console.WriteLine("Object(0x{0:X04}).DataType = {1}", controlWordOd.Index, objectEntry.getDataType().ToString());
            Console.WriteLine();

            Console.WriteLine("Some ObjectSubEntry properties:");
            Nlc.ObjectSubEntry objectSubEntry = nanolibHelper.GetObject(objectDictionary, controlWordOd);
            Console.WriteLine("OdIndex(0x{0:X04}:{1:X02}).DataType = {2}", controlWordOd.Index, controlWordOd.SubIndex, objectSubEntry.getDataType().ToString());
            Console.WriteLine("OdIndex(0x{0:X04}:{1:X02}).BitLength = {2}", controlWordOd.Index, controlWordOd.SubIndex, objectSubEntry.getBitLength().ToString());
            Console.WriteLine();

        }

        // Wait for any key so console window will not close immediately
        private static void PressAnyKey()
        {
            Console.WriteLine("Press any key...");
            Console.ReadKey();
        }

        private static bool AskTheUser(string question)
        {
            Console.Write(question);
            Console.Write(" Press 'Y'es / 'N'o: ");

            char key = Console.ReadKey().KeyChar;

            Console.WriteLine();

            return (key == 'Y') || (key == 'y');
        }
    }
}

File: CAN_DRIVER/NANOTEC/NanoLib_1.0.1/NanoLib Python Windows/nanotec_nanolib_win-1.0.1/example/nanolib_example.py
# -*- coding: utf-8 -*-

from nanolib_helper import *
from nanolib_profinet_example import ProfinetExample
from nanolib_sampler_example import SamplerExample

def askUser(question):
    userInput = input("" + str(question) + " " + "[Y/N]:" + " ")
    userInput = str(userInput).strip().upper()
    if(userInput == "Y"):
        return True
    else:
        return False

def object_code_to_string(object_code):
    return {
        Nanolib.ObjectCode_Null: 'Null',
        Nanolib.ObjectCode_Deftype: 'Deftype',
        Nanolib.ObjectCode_Defstruct: 'Defstruct',
        Nanolib.ObjectCode_Var: 'Var',
        Nanolib.ObjectCode_Array: 'Array',
        Nanolib.ObjectCode_Record: 'Record'
        }.get(object_code, 'Unknown')


def object_dictionary_access_examples(nanolib_helper, device_handle):
    print('\nOD Example\n')

    print("Reading subindex 0 of index 0x6040")
    status_word = nanolib_helper.read_number(device_handle, Nanolib.OdIndex(0x6040, 0x00))
    print('Result: {}\n'.format(status_word))

    print('Motor Stop (0x6040-0)')
    status_word = nanolib_helper.write_number(device_handle, 6, Nanolib.OdIndex(0x6040, 0x00), 16)

    print('\nRead Nanotec home page string')
    home_page = nanolib_helper.read_string(device_handle, Nanolib.OdIndex(0x6505, 0x00))
    print('The home page of Nanotec Electronic GmbH & Co. KG is: {}'.format(home_page))

    print('\nRead device error stack')
    error_stack = nanolib_helper.read_array(device_handle, Nanolib.OdIndex(0x1003, 0x00))
    print('The error stack has {} elements\n'.format(error_stack[0]))

def object_dictionary_access_examples_via_dictionary_interface(nanolib_helper, device_handle):
    home_page_od = Nanolib.OdIndex(0x6505, 0x00);
    control_word_od = Nanolib.OdIndex(0x6040, 0x00);

    print('\nOD Example via Dictionary Interface\n')

    object_dictionary = nanolib_helper.get_device_object_dictionary(device_handle)

    print("Reading subindex 0 of index 0x6040")
    status_word = nanolib_helper.read_number_od(object_dictionary, Nanolib.OdIndex(0x6040, 0x00))
    print('Result: {}\n'.format(status_word))

    print('Motor Stop (0x6040-0)')
    status_word = nanolib_helper.write_number_od(object_dictionary, 6, Nanolib.OdIndex(0x6040, 0x00))

    print('\nRead Nanotec home page string')
    home_page = nanolib_helper.read_string_od(object_dictionary, home_page_od)
    print('The home page of Nanotec Electronic GmbH & Co. KG is: {}'.format(home_page))
    
    object_entry = nanolib_helper.get_object_entry(object_dictionary, control_word_od.getIndex())
    print('Some ObjectEntry properties:')
    print('Object(0x%04X).ObjectCode = %s' % (control_word_od.getIndex(), object_code_to_string(object_entry.getObjectCode())))
    print('Object(0x%04X).DataType = %s' % (control_word_od.getIndex(), Nanolib.OdTypesHelper.objectEntryDataTypeToString(object_entry.getDataType())))

    print('Some ObjectSubEntry properties:')
    object_sub_entry = nanolib_helper.get_object(object_dictionary, control_word_od);
    print('OdIndex(0x%04X:%02X).DataType = %s' % (control_word_od.getIndex(), control_word_od.getSubIndex(), Nanolib.OdTypesHelper.objectEntryDataTypeToString(object_sub_entry.getDataType())))
    print('OdIndex(0x%04X:%02X).BitLength = %d' %(control_word_od.getIndex(), control_word_od.getSubIndex(), object_sub_entry.getBitLength()))


if __name__ == '__main__':
    nanolib_helper = NanolibHelper()

    # create access to the nanolib
    nanolib_helper.setup()
    
    print('Nanolib Example')

    # its possible to set the logging level to a different level
    nanolib_helper.set_logging_level(Nanolib.LogLevel_Off)

    # list all hardware available, decide for the first one
    bus_hardware_ids = nanolib_helper.get_bus_hardware()

    if bus_hardware_ids.empty():
        raise Exception('No bus hardware found.')
        
    print('\nAvailable bus hardware:\n')

    line_num = 0
    # just for better overview: print out available hardware
    for bus_hardware_id in bus_hardware_ids:
        print('{}. {} with protocol: {}'.format(line_num, bus_hardware_id.getName(), bus_hardware_id.getProtocol()))
        line_num += 1

    print('\nPlease select (type) bus hardware number and press [ENTER]:', end ='');
    
    line_num = int(input())
    
    print('');
    
    if ((line_num < 0) or (line_num >= bus_hardware_ids.size())):
        raise Exception('Invalid selection!')
        
    # Use the selected bus hardware
    bus_hw_id = bus_hardware_ids[line_num]

    # create bus hardware options for opening the hardware
    bus_hw_options = nanolib_helper.create_bus_hardware_options(bus_hw_id)

    # now able to open the hardware itself
    nanolib_helper.open_bus_hardware(bus_hw_id, bus_hw_options)
    
    # either scan the whole bus for devices (in case the bus supports scanning)
    device_ids = nanolib_helper.scan_bus(bus_hw_id)
    
    print("")
    for device_id in device_ids:
        print("Found Device: {}".format(device_id.toString()))
        
    if (device_ids.size() == 0):
        raise Exception('No devices found.')

    print('\nAvailable devices:\n')

    line_num = 0
    # just for better overview: print out available hardware
    for id in device_ids:
        print('{}. {} [device id: {}, hardware: {}]'.format(line_num, id.getDescription(), id.getDeviceId(), id.getBusHardwareId().getName()))
        line_num += 1

    print('\nPlease select (enter) device number(0-{}) and press [ENTER]:'.format(line_num - 1), end ='');
    
    line_num = int(input())
    
    print('');
    
    if ((line_num < 0) or (line_num >= device_ids.size())):
        raise Exception('Invalid selection!')

    # We can create the device id manually    
    # device_id = Nanolib.DeviceId(bus_hw_id, 1, "")
	# or select first found device on the bus
    device_id = device_ids[line_num]
	
    device_handle = nanolib_helper.create_device(device_id)

    # now connect to the device
    nanolib_helper.connect_device(device_handle)

    # now ready to work with the controller, here are some examples on how to access the
    # object dictionary:
    object_dictionary_access_examples(nanolib_helper, device_handle)
    object_dictionary_access_examples_via_dictionary_interface(nanolib_helper, device_handle)

    if(bus_hw_id.getBusHardware() == Nanolib.BUS_HARDWARE_ID_NETWORK):
        profinet_dcp_interface = nanolib_helper.get_profinet_dcp_interface()
        if(
            profinet_dcp_interface.isServiceAvailable(bus_hw_id).hasError() == False
        ):
            if(askUser("Do you wish to proceed with Profinet DCP example?")):
                profinetExample = ProfinetExample(bus_hw_id, profinet_dcp_interface)
                profinetExample.execute()
        
    if(askUser("Do you wish to proceed with sampler examples?")):
        samplerExample = SamplerExample(device_handle, nanolib_helper)
        print("")
        samplerExample.executeWithoutCallback()
        samplerExample.executeWithCallback()

    # cleanup and close everything
    nanolib_helper.disconnect_device(device_handle)
    nanolib_helper.close_bus_hardware(bus_hw_id)

    print("Closing everything successfully")
